# <font color='red'> <2022 Midterm Exam>  </font>
You have  until **2:15pm** to finish the exam. 

You are <font color='red'>**NOT**</font> allowed to go through our <font color='red'> **class material files**, **lecture slides**, or **google** </font>. <br>
If you get caught, then you will be given 0 for your entire final exam score, including LMS **WITHOUT ANY EXCEPTION.**

You have to UPLOAD your exam by **2:15pm**. I will be closing the upload at 2:15pm sharp.<br>
If you are done, save it and upload it on LMS. <br>
If you have any question, please ask me! <br>

Please make sure that you READ and FOLLOW the given instructions carefully. <br>
eg) if I ask you to create a variable named <code>abc</code>, then please create a variable with the EXACT name.<br>
If you create a variable as <code>ABC</code> or <code>cde</code> then I will considered to have not followed the written instructions. <br>
Therefore, marks will be deducted.
<br><br>
If you have any questions, please raise your hand<br>
Good luck!

# <font color='red'> Question 1: 2 points

I would like to import pandas library with the alias pd. Write one line code that would do so.

In [1]:
import pandas as pd

# <font color='red'> Question 2: 2 points

Change the following function using lambda function. <br>
<code>def square(x):
    return x*x

In [2]:
square = (lambda x: x * x)

# <font color='red'> Question 3: 3 points

I ran the following code. <br>
<code>alice_contents=[word.lower() for word in alice_contents if word.lower() not in stopwords]</code>

Explain in words what this line of code would do. <br>
You should explain in terms of the following key words. (1 point for each)<br>
Point 1. For loop
Point 2. If
Point 3. Word.lower
Point 4. Alice contents


write your answer in this box in your own words:
"alice_contents" is a variable, probably a list. "alice_contents" has to be an iterable, because we are using the for-each-loop on it.
A for-each-loop iterates through an iterable and execute code for each item in the iterable. So for every item in the list "alice_contents" we execute code, in our case we execute "lower()".
The "word.lower()" can be executed on strings and it transforms the string into all lowercase letters.
We also have a part "if word.lower() not in stopwords". This part says that the execution of code on the left is only done if the if expression is True. So only if the lowercase version of the word in not in stopwords.

The whole for loop then returns all the items in the list which if was true.

# <font color='red'> Question 4: 4 points

In [3]:
# run the following code
import requests
from bs4 import BeautifulSoup

page = requests.get("http://quotes.toscrape.com/")
soup = BeautifulSoup(page.content, 'html.parser')
print(soup.prettify())

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   Quotes to Scrape
  </title>
  <link href="/static/bootstrap.min.css" rel="stylesheet"/>
  <link href="/static/main.css" rel="stylesheet"/>
 </head>
 <body>
  <div class="container">
   <div class="row header-box">
    <div class="col-md-8">
     <h1>
      <a href="/" style="text-decoration: none">
       Quotes to Scrape
      </a>
     </h1>
    </div>
    <div class="col-md-4">
     <p>
      <a href="/login">
       Login
      </a>
     </p>
    </div>
   </div>
   <div class="row">
    <div class="col-md-8">
     <div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
      <span class="text" itemprop="text">
       “The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
      </span>
      <span>
       by
       <small class="author" itemprop="author">
        Albert Einstein
       </small>
       <a href="/author/Albert

<font color='red'>**WITHOUT** connecting to the website,</font><br>meaning just by using what has been printed using by soup.prettify() above,<br>
Find all 'quotes' (Not authors,not links)  from this page and **save your quotes in a variable named <code>text</code>**

In [4]:
text = soup.find_all("span", class_="text")

# <font color='red'> Question 5: 4 points

I want to get the text out from <code>text</code> variable that we just made in Q4. <br><BR>
To do that, we need to loop through the text variable and tell Python to: <br>
1. go through text and for each element in a list
2. we will get text out from it. <br>
3. Then, extract text and save them in a variable called <code>quotes</code>.<br>

**hint!** <br>
To do so, we will first need to make an empty variable called <code>quotes</code>.<br>
Then, using loops again, we will <code>append</code> them in the variable.

In [5]:
quotes = []
for t in text:
    quotes.append(t.get_text())

# <font color='red'> Question 6: 3 points

Save your data in a dataframe called <code>quotes</code> with column name for quotes being "quotes"

In [6]:
quotes = pd.DataFrame(data=quotes, columns=["quotes"])

# <font color='red'> Question 7: 3 points

In [7]:
#run the following code
import requests
import nltk
import re
from nltk.corpus import gutenberg

alice = gutenberg.raw(fileids='carroll-alice.txt')
stopwords = nltk.corpus.stopwords.words('english')

After you run the above code, <br>
Tokenize **<code>alice</code>** variable into SENTENCES and save tokenized text into 
**<code>alice_sentences</code>**

In [8]:
alice_sentences = nltk.sent_tokenize(alice)

# <font color='red'> Question 8: 4 points

Replace all "\n" with an empty string for <code>alice_sentences</code>, (3 points)<br>
and save it back into **<code>alice_sentences</code>**(1 point)

In [9]:
alice_sentences = [re.sub("\n", " ", sentence) for sentence in alice_sentences] # empty string or string with only a whitespace?

# <font color='red'> Question 9: 3 points
   

Using <code>alice</code> variable from Q7, <br>
Tokenize into WORDS this time, and save it in a variable called <code>alice_word</code>.

In [10]:
alice_word = nltk.word_tokenize(alice)

# <font color='red'> Question 10: 4 points.

Explain in your own words, what happens when I run the following code. <br>
Try explaining step by step on what happens in the code below. <br><br>
**Hint** <br>
Regex pattern [^A-Za-z] means <br> "matches a single character not present in the range between A-Z and a-z" <br>
meaning, it matches non alphabet, upper and lower cases. <br>

In [11]:
alice_contents = list(filter(None, [re.sub(r'[^A-Za-z]', '', a) for a in alice_word]))

In [12]:
# explain in your words here:
# The re.sub(l,m,n) function finds everything in n (in our case variable a) that matches the regular expression l (in our case [^A-Za-z]) and replaces it with m (in our case '').
# The for loop iterates through alice_word, so the re.sub() is called for every word in alice_word.
# the filter-function removes every element that is None in the second argument (so our result after the for-loop-re.sub-thing).
# the list-function makes a list out of the result of the filter function
# the result of the list-function is then saved in the variable alice_contents

# <font color='red'> Question 11: 4 points.

In [13]:
#run the following code
alice_contents = [word.lower() for word in alice_contents if word.lower() not in stopwords]

Now that we changed into lower cases for all alice_contents,<br>
I want to get rid of the tokens that contain stopwords. <br>
1. iterate through <code>alice_contents</code>
2. **if** each of the iterable token **does not** contain stopwords,
3. save it back into <code>alice_contents</code>

This means, for each word in alice_contents, get rid of stopwords and save it back into <code>alice_contents</code>

In [14]:
alice_contents = [content for content in alice_contents if content not in stopwords]
alice_contents

['alice',
 'adventures',
 'wonderland',
 'lewis',
 'carroll',
 'chapter',
 'rabbithole',
 'alice',
 'beginning',
 'get',
 'tired',
 'sitting',
 'sister',
 'bank',
 'nothing',
 'twice',
 'peeped',
 'book',
 'sister',
 'reading',
 'pictures',
 'conversations',
 'use',
 'book',
 'thought',
 'alice',
 'without',
 'pictures',
 'conversation',
 'considering',
 'mind',
 'well',
 'could',
 'hot',
 'day',
 'made',
 'feel',
 'sleepy',
 'stupid',
 'whether',
 'pleasure',
 'making',
 'daisychain',
 'would',
 'worth',
 'trouble',
 'getting',
 'picking',
 'daisies',
 'suddenly',
 'white',
 'rabbit',
 'pink',
 'eyes',
 'ran',
 'close',
 'nothing',
 'remarkable',
 'alice',
 'think',
 'much',
 'way',
 'hear',
 'rabbit',
 'say',
 'oh',
 'dear',
 'oh',
 'dear',
 'shall',
 'late',
 'thought',
 'afterwards',
 'occurred',
 'ought',
 'wondered',
 'time',
 'seemed',
 'quite',
 'natural',
 'rabbit',
 'actually',
 'took',
 'watch',
 'waistcoatpocket',
 'looked',
 'hurried',
 'alice',
 'started',
 'feet',
 'flas

# <font color='red'> Question 12: 2 points.

Conduct pos tagging of <code>alice_contents</code> (1 point) <br>
and save the tagged tokens into a variable called <code>alice_pos</code> (1 point) 

In [15]:
alice_pos = nltk.pos_tag(alice_contents)

# <font color='red'> Question 13: 2 points

Change the following code using list comprehension

<code>nouns=[]
for word, pos in alice_pos:
    if pos in ['NN']:
        nouns.append(word)</code>


In [16]:
nouns = [word for word, pos in alice_pos if pos in ['NN']]
nouns

['alice',
 'carroll',
 'chapter',
 'alice',
 'bank',
 'nothing',
 'book',
 'sister',
 'reading',
 'book',
 'thought',
 'alice',
 'mind',
 'well',
 'day',
 'feel',
 'pleasure',
 'making',
 'daisychain',
 'trouble',
 'rabbit',
 'nothing',
 'alice',
 'way',
 'rabbit',
 'time',
 'rabbit',
 'watch',
 'waistcoatpocket',
 'alice',
 'mind',
 'rabbit',
 'waistcoatpocket',
 'watch',
 'burning',
 'curiosity',
 'field',
 'time',
 'rabbithole',
 'hedge',
 'moment',
 'world',
 'get',
 'way',
 'moment',
 'time',
 'wonder',
 'look',
 'dark',
 'anything',
 'jar',
 'orange',
 'disappointment',
 'drop',
 'jar',
 'fear',
 'somebody',
 'alice',
 'fall',
 'nothing',
 'home',
 'anything',
 'house',
 'end',
 'wonder',
 'time',
 'earth',
 'let',
 'sort',
 'schoolroom',
 'opportunity',
 'listen',
 'practice',
 'distance',
 'wonder',
 'latitude',
 'longitude',
 'idea',
 'latitude',
 'earth',
 'listening',
 'time',
 'sound',
 'word',
 'country',
 'zealand',
 'spoke',
 'fancy',
 'air',
 'think',
 'girl',
 'nothing

# <font color='red'> Question 14: 4 points

In [17]:
#run the following code
from pygooglenews import GoogleNews

gn = GoogleNews()


def get_titles(search):
    """fetch news titles"""
    stories = []
    search = gn.search(search)
    newsitems = search['entries']
    for item in newsitems:
        story = {
            'title': item.title,
        }
        stories.append(story)
    return stories


title = get_titles('midterm exam')
title = pd.DataFrame(title)
title

title
0   OPINION: Midterm and final exams are unnecessa...
1   Stanford's Social Life 101 deemed most difficu...
2   Arguing With Myself About Whether To Call Midt...
3   Could NAEP affect midterm elections? And other...
4   Mid-term elections put both parties to the tes...
..                                                ...
95  Stamford high schools suspend midterms, finals...
96  In-person midterms are preferred over online t...
97  Midterm Exam Coming Up? You Must Do This One T...
98  F.D.A. Clears Path for Over-the-Counter Hearin...
99  Post-secondary life could be tough test for pa...

[100 rows x 1 columns]

After importing GoogleNews, I ran the following code above. <br>
Then, I managed to get the list of news articles with 'midterm exam' as the keywords. <br>
I notice that there are both title and sources within our 'title' column. <br>
Therefore, I want to retrieve only the title without sources. <br><br>
*note* <br>
Title and sources can be separated by '-'<br>
Use title dataframe to work on it. <br><br>
Using the str.split function, <br>
1. create a new column called 'Title'. and within that newly created column,
2. using the str.split() function, split 'title' column by the separator '-' <br>
3. Set expand=True
4. Make sure you use the first half before the separator, which is denoted by **[0]** (hint: add [0] at the end...)

In [18]:
title['Title'] = title['title'].str.split("-", expand=True)[0]
title

title  \
0   OPINION: Midterm and final exams are unnecessa...   
1   Stanford's Social Life 101 deemed most difficu...   
2   Arguing With Myself About Whether To Call Midt...   
3   Could NAEP affect midterm elections? And other...   
4   Mid-term elections put both parties to the tes...   
..                                                ...   
95  Stamford high schools suspend midterms, finals...   
96  In-person midterms are preferred over online t...   
97  Midterm Exam Coming Up? You Must Do This One T...   
98  F.D.A. Clears Path for Over-the-Counter Hearin...   
99  Post-secondary life could be tough test for pa...   

                                                Title  
0   OPINION: Midterm and final exams are unnecessa...  
1   Stanford's Social Life 101 deemed most difficu...  
2   Arguing With Myself About Whether To Call Midt...  
3   Could NAEP affect midterm elections? And other...  
4                                                 Mid  
..                                                ...  
95  Stamford high schools suspend midterms, finals...  
96                                                 In  
97  Midterm Exam Coming Up? You Must Do This One T...  
98                        F.D.A. Clears Path for Over  
99                                               Post  

[100 rows x 2 columns]

# <font color='red'> Question 15: 3 points

In [19]:
#run the following code
from textblob import Word, TextBlob

text = TextBlob("""
I really hated exams. I tried to make it not too difficult, but we will see.
Do not worry too much, I will bunk the grades up as possible,
so please do not give up.
""").sentiment
text

Sentiment(polarity=-0.3, subjectivity=0.725)

Explain what polarity and subjectivity number refers to.<br>
Explain within 4 sentences.

explain in this box:
In the sentiment analysis we look at to different things: polarity and subjectivity.
Polarity is whether the text is negative, positive or neutral. We measure this from -1 to 1, with one being positive, 0 being neutral and -1 being negative.
Subjectivity is the amount of attitude or emotion or personal opinion in the text. We measure this from 0 to 1, with 1 being totally subjective.
In our case the polarity is -0.3 so the text is considered to be negative, and the subjectivity is 0.725, so the text is considered subjective.

# <font color='red'> Question 16: 3 points

(Points are given for this question) <br>

**From Prof.:** <br>
Any comments of the exam, or the class, <br>
so that I can improve on the latter half of the course?? <br>
Was the exam too hard, okayish or easy? <br>
If you have any comments, please leave it here. 

In [20]:
# I think it was appropriate. I studied for the exam, did the exercises and so for me it was fine. I think if you haven't studied it might be hard, which is fair.
# Before the exam I was worried because I thought more exercise on my own would have helped me more, but now that I did the exam I think the exercise that we did helped me more then enough.

#  <font color='GREEN'> ONCE YOU ARE DONE, SAVE YOUR FILE WITH YOUR NAME, AND UPLOAD IN THE EXAM SECTION.